### ISSUE TITLE
DeepSeek-V4: llama_kv_cache_dsv4 state does not round-trip through state_seq_get/set (kv_state_spike 3/6 fail)

### ISSUE THREAD

**Issue by fpt** [fpt/rs-gallium#209]
## The problem

`crates/gallium-agent/tests/kv_state_spike.rs` run against DeepSeek-V4
(`unsloth/DeepSeek-V4-Flash-0731-GGUF/UD-IQ3_XXS`, `GALLIUM_KV_SPIKE_CPU_MOE=1`,
RTX 4070) fails all three equivalence tests. Deterministic across two runs to the
last digit:

```
test result: FAILED. 3 passed; 3 failed
```

| test | Δ top-1 logit (tolerance `LOGIT_TOLERANCE = 0.1`) |
|---|---|
| `a_restore_with_nothing_in_between_is_the_same_state` | **1.689527** — snapshot then restore with *no generation in between*, 1681 tokens |
| `a_sequence_snapshot_..._costs_less_than_a_prefill` | **0.808103** — 12-token greedy continuation agrees, logits do not |
| `an_on_device_checkpoint_restores_an_equivalent_state` | **1.689527** — identical to the host-snapshot path, so `ON_DEVICE` is no way around it |

For comparison the same test is **0.000000** on LFM2.5 and **0.021921** (FP
reordering from cell placement) on Gemma 4 E4B. The 3 "passed" are the two
informational tests (`what_each_state_flag_costs`, `where_the_first_get_spends_its_time`)
and `a_recurrent_only_snapshot_...`, which SKIPs for a non-LFM2 model.

`a_restore_with_nothing_in_between` is the clean one: it calls `state_seq_get`
then `state_seq_set` immediately, with nothing between, and compares against a
context that never snapshotted. Anything non-zero is the get/set cycle's own
doing, and 1.69 logits is neither cell placement nor noise. `llama_kv_cache_dsv4`
(`kv_raw` plus three compressed sub-caches — `p0/DSV4_CSA_RATIO`,
`p0/DSV4_HCA_RATIO`, the lightning-indexer — addressed at different position
scales) does not restore consistently.

## What is ruled out

- **The cost side is fine.** Snapshot 28.3 MiB, `state_seq_get` 7.2 ms /
  `state_seq_set` 4.2 ms, restore+suffix = 0.1x the prefill; on-device handle
  only 27 KiB. It is *equivalence* that fails, not performance.
- **Not the 2026-08-28 `file_read` / `data_analysis` matrix failures.** Those
  were `reasoningEffort` unset (the model was prompted to act with no
  scratchpad), fixed by `reasoningEffort = "medium"`, and reproduced identically
  with `GALLIUM_KV_CACHE_SLOTS=0` (checkpoint code fully bypassed). See the
  DeepSeek-V4-Flash section of `docs/VERIFICATION_STATUS.md`. The checkpoint has
  simply not been *caught* changing a testcase outcome — a 1.69 Δlogit that
  happens to keep the same argmax on the cases run so far.

## Why it matters

A 1.69 Δlogit on a bare state restore is exactly the silent-corruption shape
#196's gate exists to prevent. The gate currently latches on an observed
partial-rollback refusal (`llama_kv_cache_dsv4::seq_rm` refuses every real one)
and then uses `Slot::checkpoint` — `state_seq_get`/`set` — to roll back instead of
re-evaluating the transcript. On this cache that restore is wrong, so a
`deepseek4` model in a multi-iteration ReAct turn is running from a KV state that
is close but not equal, with only a `debug!` line to say a checkpoint was used.

## Candidates for whoever picks this up

- Whether `state_seq_get`/`set` in the vendored llama.cpp actually serialise and
  restore all three dsv4 compressed sub-caches, or only `kv_raw` — an
  inconsistent partial restore across the sub-caches would produce exactly this
  "plausible but wrong" shape.
- Whether the gate should treat a `deepseek4` refusal as "re-evaluate the
  transcript" rather than "restore the checkpoint" until the round-trip is fixed
  — i.e. an arch/cache denylist for checkpoints, distinct from the reuse gate.
- Whether it reproduces at full GPU offload (this run used `cpuMoe`) and at a
  higher quant than `UD-IQ3_XXS`.
- An upstream llama.cpp check: does `llama_kv_cache_dsv4`'s `state_write` /
  `state_read` cover the compressed states, and is there a known issue there.

## Reference

- Test: `crates/gallium-agent/tests/kv_state_spike.rs` (its module doc has the
  exact command)
- Docs: `docs/VERIFICATION_STATUS.md` (DeepSeek-V4-Flash section), `docs/OPTIMIZATION.md` §3.5
- Landed in: #207
- Related: #196 (the checkpoint mechanism and its gate)

Repro:

```
GALLIUM_KV_SPIKE_MODEL=unsloth/DeepSeek-V4-Flash-0731-GGUF/UD-IQ3_XXS/DeepSeek-V4-Flash-0731-UD-IQ3_XXS-00001-of-00004.gguf \
GALLIUM_KV_SPIKE_CPU_MOE=1 \
  cargo test --release --features cuda -p gallium-agent --test kv_state_spike -- --ignored --nocapture --test-threads=1
```

**Comment by fpt** [comment]
**Safe half done in #211.** `arch_checkpoint_state_round_trips(arch)` in `llm_local.rs` denylists `deepseek4` from `Slot::checkpoint` (distinct from the reuse gate — the gate still latches on the observed refusal, it just no longer rolls back via `state_seq_set`). A `deepseek4` refused partial trim now re-evaluates the transcript: slower, but never running from an almost-equal KV state. Unit-tested model-free (`only_deepseek4_disables_state_checkpoints`).

Re-confirmed the repro on the CUDA box (UD-IQ3_XXS, `GALLIUM_KV_SPIKE_CPU_MOE=1`, `--features cuda`) before making the change — `test result: FAILED. 3 passed; 3 failed`, same split as the original, 253s.

**Still open (the hard half):** whether `llama_kv_cache_dsv4`'s `state_write`/`state_read` in the vendored llama.cpp actually serialise all three compressed sub-caches (`csa` 1.31 MiB, `hca` 10.00 MiB, lightning-indexer `lid` 0.33 MiB — sizes from the load log) or only `kv_raw`, and the upstream check on the same. Until that lands, `kv_state_spike` fails 3/3 on this model and #209 stays open.
